domain/modelagent/service/service.go

```diff
@@ -499,6 +579,356 @@ func (s *Service) SetUnitReportedAgentVersion(
+	err := s.UpgradeModelTargetAgentVersionStreamTo(
+		ctx, recommendedVersion, agentStream,
+	)
+
+	// NOTE (tlm): Because this func uses
+	// [Service.UpgradeModelTargetAgentVersionStreamTo] to compose it's
+	// implementation. This func must handle the contract of
+	// UpgradeModelTargetAgentVersion. Specifically the errors returned don't
+	// align with the expecations of the caller. The below switch statement
+	// re-writes the error cases to better explain the very unlikely error that
+	// has occurred. These exists to point a developer at the problem and not to
+	// offer any actionable item for a caller.
+	switch {
+	case errors.Is(err, modelagenterrors.DowngradeNotSupported):
+		return semversion.Zero, errors.Errorf(
+			"upgrading model to recommended version %q is considered a downgrade of the model agent version",
+			recommendedVersion,
+		)
+	case errors.Is(err, modelagenterrors.AgentVersionNotSupported):
+		return semversion.Zero, errors.Errorf(
+			"upgrading model to recommended version %q is not a supported version number",
+			recommendedVersion,
+		)
+	case errors.Is(err, modelagenterrors.MissingAgentBinaries):
+		return semversion.Zero, errors.Errorf(
+			"upgrading model to recommended version %q is missing agent binaries for version",
+			recommendedVersion,
+		)
+	case err != nil:
+		return semversion.Zero, errors.Errorf(
+			"upgrading model to recommended version %q: %w",
+			recommendedVersion, err,
+		)
+	}
+
+	return recommendedVersion, nil
+}
+
+// UpgradeModelTargetAgentVersionTo upgrades a model to a new target agent
+// version. All agents that run on behalf of entities within the model will be
+// eventually upgraded to the new version after this call successfully returns.
+//
+// The version supplied must not be a downgrade from the current target agent
+// version of the model. It must also not be greater then the maximum supported
+// version of the controller.
+//
+// The following errors may be expected:
+// - [modelagenterrors.DowngradeNotSupported] when the caller is attempting to
+// change the model target agent version to one that is lower than the current
+// version.
+// - [modelagenterrors.AgentVersionNotSupported] when the caller is attempting
+// to upgrade the version to one that is greater then the maximum supported
+// version of the controller hosting the model.
+// - [modelagenterrors.MissingAgentBinaries] when agent binaries do not exist
+// for the considered version. The model cannot be upgraded to a version where
+// no agent binaries exist.
+// - [modelagenterrors.CannotUpgradeControllerModel] when the current model is
+// the model running the Juju controller.
+// - [modelagenterrors.ModelUpgradeBlocker] when their exists a blocker in the
+// model that prevents the model from being upgraded.
+func (s *Service) UpgradeModelTargetAgentVersionTo(
+	ctx context.Context,
+	consideredVersion semversion.Number,
+) error {
+	ctx, span := trace.Start(ctx, trace.NameFromFunc())
```

Comment:
When called by `UpgradeModelTargetAgentVersion` which has already started a Span, there's no point doing it again. There should be a private helper called by both `UpgradeModelTargetAgentVersion` and `UpgradeModelTargetAgentVersionTo`

Reply:
I am not sure it makes a difference in this case. It is just more context added to the span. Performing the composition dance just for the sake of a span seems overkill. You just get more information then you bargained for.

Reply:
Won't block on it but I'd still prefer it for "cleanliness".